File: Group Anagrams/groupAnagrams.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        # create a dictionary to group anagrams
        anagrams = defaultdict(list) 

        for word in strs:
            # sort the word and use that as the key
            key = ''.join(sorted(word))
            anagrams[key].append(word) #add the word to the dictionary using the key that is the sorted alphabet

        return list(anagrams.values()) # return list of grouped anagrsms from the dictionary  basically all values in dict except keys

File: Group Anagrams/groupAnagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        // create hash map
        unordered_map<string, vector<string>> anagrams;
        for ( const string& word : strs) {
            //sort to generate key
            string key = word;
            sort(key.begin(), key.end());

            // add word to proper key
            anagrams[key].push_back(word);

        }
        // collect the grouped anagrams into a vector of vectors to return it
        vector<vector<string>> result;
        for ( const auto& group : anagrams){
            result.push_back(group.second); // second since the first would be the key
        }
        return result;
    }
};
